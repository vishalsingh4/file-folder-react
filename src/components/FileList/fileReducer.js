import { getFolderNameFromFileName } from "../../utils";

const initialState = {
    currentFileItem: {},
    inputValue: ''
};

export const getDefaultInitialState = ({
    defaultCurrentFileItem = {}, 
    defaultInputValue = ''
}) => {
    return {
        ...initialState,
        currentFileItem: defaultCurrentFileItem,
        inputValue: defaultInputValue
    }
}

export const FILE_ACTIONS = {
    INPUT_CHANGE: 'INPUT_CHANGE',
    TOGGLE_ICON_CLICK: 'TOGGLE_ICON_CLICK',
    ADD_FILE: 'ADD_FILE',
    DBL_CLICK_FILE: 'DBL_CLICK_FILE'
};

export default function fileReducer(state = initialState, action) {
    const {
        type,
        payload
    } = action;

    switch(type) {
        case FILE_ACTIONS.INPUT_CHANGE: {
            return {
                ...state,
                inputValue: payload
            }
        }
        case FILE_ACTIONS.TOGGLE_ICON_CLICK: {
            return {
                ...state,
                currentFileItem: {
                    ...state.currentFileItem,
                    isOpen: !state.currentFileItem.isOpen
                }
            }
        }
        case FILE_ACTIONS.ADD_FILE: {
            const hasFilesProperty = state.currentFileItem.hasOwnProperty('files');
            debugger;
                    if (hasFilesProperty) {
                        return {
                            ...state,
                            currentFileItem: {
                                ...state.currentFileItem,
                                files: [payload, ...state.currentFileItem.files]
                            }
                        }
          } else {
            return {
                ...state,
                currentFileItem: {
                    ...state.currentFileItem,
                    files: [payload]
                }
            }
          }
        }
        case FILE_ACTIONS.DBL_CLICK_FILE: {
            return {
                ...state,
                currentFileItem: {
                    ...state.currentFileItem,
                    name: getFolderNameFromFileName(payload),
                    isFolder: true
                }
            }
        }
        default:
            return state;
    }
};