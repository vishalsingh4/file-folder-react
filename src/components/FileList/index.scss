@import '../../assets/base.scss';

.file-item {
    >button {
        background-color: $green;
        color: $white;
        padding: $spacing-l;
        border-radius: 4px;
        border: 1px solid $green;
        margin-bottom: $spacing-xs;
        display: flex;
        justify-content: space-between;

        .name {
            max-width: 150px;
            overflow: hidden;
            text-overflow: ellipsis;
            white-space: nowrap;
        }

        .toggle-icon {
            margin-left: $spacing-l;
            cursor: pointer;
        }
    }
}

.file-input {
    padding: $spacing-xs $spacing-l;
    margin-bottom: $spacing-s;
    border-radius: 4px;
    border: 1px solid $black;
}

.nested-files {
    padding-left: $spacing-xl;
}