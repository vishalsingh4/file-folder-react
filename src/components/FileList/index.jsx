import React, { Fragment, useReducer } from 'react';
import { v4 as uuidv4 } from 'uuid';

import Button from '../Button';
import Input from '../Input';

import { isValidFileName, validateFileInput } from '../../utils';
import fileReducer, { getDefaultInitialState, FILE_ACTIONS } from './fileReducer';

import './index.scss';

function FileList({
  fileItem = {},
}) {
  const [state, dispatch] = useReducer(fileReducer, getDefaultInitialState({
    defaultCurrentFileItem: fileItem
  }));

  const {
    currentFileItem,
    inputValue
  } = state;

  const {
    id,
    name,
    isOpen,
    files,
    isFolder
  } = currentFileItem;

  const isNodeModulesFolder = name === 'node_modules';

  const handleToggle = (e) => {
    e.preventDefault();

    dispatch({
      type: FILE_ACTIONS.TOGGLE_ICON_CLICK
    })
  };

  const handleAddFile = (e) => {
    const enterBtnPressed = (e.key === 'Enter' || e.keyCode === 13);

    if (enterBtnPressed) {
      const userInput = e.target.value;
      const isValid = validateFileInput(currentFileItem, userInput);

      if (!isValid) {
        return false;
      }

      const newItem = {
        name: userInput,
        id: uuidv4(),
        isFolder: false
      }

      dispatch({
        type: FILE_ACTIONS.ADD_FILE,
        payload: newItem
      })

      dispatch({
        type: FILE_ACTIONS.INPUT_CHANGE,
        payload: ''
      })
    }
  };

  const handleDblClick = (e) => {
    const selectedTextName = e.target.dataset.fileName;
    const isFileName = isValidFileName(selectedTextName);

    if (!isFileName) {
      return false;
    }

    dispatch({
      type: FILE_ACTIONS.DBL_CLICK_FILE,
      payload: selectedTextName
    })
  };

  return (
    <>
      <li key={id} className='file-item'>
        <Button
          name={name}
          onDblClick={handleDblClick}
          showIcon={isFolder && !isNodeModulesFolder}
          onToggleIconClick={handleToggle}
          showOpenIcon={isOpen}
        />
      </li>
      {isOpen && (
        <>
          <Input
            value={inputValue}
            onChange={e => {
              dispatch({
                type: FILE_ACTIONS.INPUT_CHANGE,
                payload: e.target.value
              })
            }
            }
            onKeyDown={handleAddFile}
            data-testid="input-box"
            className='file-input'
            placeholder='Please enter file name'
            autoFocus
            required
          />
          {
            files?.length > 0 && (
              <div className='nested-files'>
                {
                  files?.map(item => (
                    <Fragment key={uuidv4()}> <FileList fileItem={item} /></Fragment>
                  ))
                }
              </div>
            )
          }
        </>
      )}
    </>
  )
}

export default FileList;