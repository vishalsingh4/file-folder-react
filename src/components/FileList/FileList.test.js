import { render, cleanup } from '@testing-library/react';

import FileList from './index';

describe('<FileList />', () => {
    afterEach(cleanup);

    it('should match snapshot', () => {
        const { asFragment } = render(<FileList />);
        expect(asFragment).toMatchSnapshot();
    });

    it('should render FileList component', () => {
        const { getByRole } = render(<FileList />);
        const button = getByRole('button');
        expect(button).toBeInTheDocument();
    })
})