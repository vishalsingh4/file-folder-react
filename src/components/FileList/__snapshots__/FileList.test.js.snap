// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileList /> should match snapshot 1`] = `[Function]`;

exports[`<FileList /> should render FileList component 1`] = `[Function]`;
